Modules/_asynciomodule.c

```diff
@@ -2713,14 +2713,19 @@ task_step_impl(TaskObj *task, PyObject *exc)
 
         if (task->task_must_cancel) {
             PyObject *r;
-            r = future_cancel(fut);
+            int is_true;
+            r = _PyObject_CallMethodId(fut, &PyId_cancel, NULL);
```

Comment:
The fix is correct.
But I'd like to go one step forward: use `Future_CheckExact` and `Task_CheckExact` to call `future_cancel()` and `_asyncio_Task_cancel_impl()` as fast paths.

Reply:
Id keep it as simple as possible as it needs to be backported to 3.6 & 3.7. A proper refactoring for C code is needed anyways, and I'll add this optimization.

Reply:
Ok